widget/color.go

```diff
@@ -0,0 +1,254 @@
+	}
+	fyne.CurrentApp().Preferences().SetString(preferenceRecents, strings.Join(recents, ","))
+}
+
+func colorToString(c color.Color) string {
+	var red, green, blue, alpha uint8
+	switch col := c.(type) {
+	case *color.NRGBA:
+		red, green, blue, alpha = col.R, col.G, col.B, col.A
+	default:
+		r, g, b, a := c.RGBA() // TODO FIXME this returns alpha-pre-multiplied
+		// TODO FIXME possible loss of precision, should this be x/65535*255?
+		red = uint8(r)
+		green = uint8(g)
+		blue = uint8(b)
+		alpha = uint8(a)
+	}
+	if alpha == 0xff {
+		return fmt.Sprintf("#%02x%02x%02x", red, green, blue)
+	}
+	return fmt.Sprintf("#%02x%02x%02x%02x", red, green, blue, alpha)
+}
+
+func stringToColor(s string) color.Color {
+	var c color.NRGBA
+	var err error
+	if len(s) == 7 {
+		c.A = 0xFF
+		_, err = fmt.Sscanf(s, "#%02x%02x%02x", &c.R, &c.G, &c.B)
+	} else {
+		_, err = fmt.Sscanf(s, "#%02x%02x%02x%02x", &c.R, &c.G, &c.B, &c.A)
+	}
+	if err != nil {
+		fyne.LogError("Couldn't parse color:", err)
+	}
+	return c
+}
+
+func stringsToColors(ss ...string) (colors []color.Color) {
+	for _, s := range ss {
+		if s == "" {
+			continue
+		}
+		colors = append(colors, stringToColor(s))
+	}
+	return
+}
+
+func colorToHSLA(c color.Color) (float64, float64, float64, float64) {
+	r, g, b, a := colorToRGBA(c)
+	h, s, l := rgbToHsl(r, g, b)
+	return h, s, l, a
+}
+
+func colorToRGBA(c color.Color) (r, g, b, a float64) {
+	switch col := c.(type) {
+	case *color.NRGBA:
+		// Convert to range 0.0-1.0
+		r = float64(col.R) / 255.0
+		g = float64(col.G) / 255.0
+		b = float64(col.B) / 255.0
+		a = float64(col.A) / 255.0
+	default:
+		red, green, blue, alpha := c.RGBA() // TODO FIXME this returns alpha-pre-multiplied
+		// Convert to range 0.0-1.0
+		r = float64(red) / 65535.0
+		g = float64(green) / 65535.0
+		b = float64(blue) / 65535.0
+		a = float64(alpha) / 65535.0
+	}
+	return
+}
+
+func cartesianToPolar(x, y, width, height float64) (angle, radius, limit float64) {
+	dx := float64(x - (width / 2))
+	dy := float64(y - (height / 2))
+	angle = math.Atan2(dy, dx)
+	radius = math.Sqrt(dx*dx + dy*dy)
+	limit = math.Min(width, height) / 2.0
+	return
+}
+
+func polarToCartesian(angle, radius, width, height float64) (x, y float64) {
+	x = (radius * math.Cos(angle)) + (width / 2)
+	y = (radius * math.Sin(angle)) + (height / 2)
+	return
+}
+
+func polarToHS(angle, radius, limit float64) (hue, saturation float64) {
+	hue = angle / (2.0 * math.Pi)
+	if hue < 0.0 {
+		hue += 1.0
+	}
+	if hue > 1.0 {
+		hue -= 1.0
+	}
+	saturation = radius / limit
+	return
+}
+
+// The next three functions are adpated from https://play.golang.org/p/9q5yBNDh3W
+
+func rgbToHsl(r, g, b float64) (float64, float64, float64) {
+	min := math.Min(r, math.Min(g, b)) //Min. value of RGB
+	max := math.Max(r, math.Max(g, b)) //Max. value of RGB
+	del := max - min                   //Delta RGB value
+
+	l := (max + min) / 2.0
+
+	var h, s float64
+
+	if del == 0.0 { // Achromatic
+		h = 0.0
+		s = 0.0
```

Comment:
It might be a bit nicer to just do this and avoid the else statement.
```suggestion
		return 0, 0, l
```

Reply:
Fixed